File: platypus-js-web-client/src/main/java/com/eas/widgets/containers/Container.java
package com.eas.widgets.containers;

import com.eas.ui.EventsPublisher;
import com.eas.ui.Widget;
import com.eas.ui.events.ContainerEvent;
import com.eas.ui.events.AddHandler;
import com.eas.ui.events.HasAddHandlers;
import com.eas.ui.events.HasRemoveHandlers;
import com.eas.ui.events.RemoveHandler;
import com.google.gwt.core.client.JavaScriptObject;
import com.google.gwt.dom.client.Style;
import com.google.gwt.event.shared.HandlerRegistration;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 *
 * @author mgainullin
 */
public abstract class Container extends Widget implements HasAddHandlers, HasRemoveHandlers {

    protected final List<Widget> children = new ArrayList<>();

    protected Container() {
        super();
        element.getStyle().setPosition(Style.Position.RELATIVE);
        element.getStyle().setOverflow(Style.Overflow.HIDDEN);
    }

    public int getCount() {
        return children.size();
    }

    public Widget get(int index) {
        if (index >= 0 && index < children.size()) {
            return children.get(index);
        } else {
            return null;
        }
    }

    public int indexOf(Widget w) {
        return children.indexOf(w);
    }

    public void add(Widget w) {
        children.add(w);
        element.appendChild(w.getElement());
        fireAdded(w);
    }

    public void add(Widget w, int beforeIndex) {
        children.add(beforeIndex, w);
        w.setParent(this);
        if (beforeIndex < children.size()) {
            element.insertBefore(w.getElement(), children.get(beforeIndex).getElement());
        } else {
            element.appendChild(w.getElement());
        }
        fireAdded(w);
    }

    public boolean remove(Widget w) {
        boolean removed = children.remove(w);
        if (removed) {
            w.setParent(null);
            w.getElement().removeFromParent();
            fireRemoved(w);
        }
        return removed;
    }

    public Widget remove(int index) {
        if (index >= 0 && index < children.size()) {
            Widget w = children.remove(index);
            w.setParent(null);
            w.getElement().removeFromParent();
            fireRemoved(w);
            return w;
        } else {
            return null;
        }
    }

    public void clear() {
        for (int i = 0; i < children.size(); i++) {
            remove(i);
        }
        children.clear();
    }

    private final Set<AddHandler> addHandlers = new HashSet<>();

    @Override
    public HandlerRegistration addAddHandler(AddHandler handler) {
        addHandlers.add(handler);
        return new HandlerRegistration() {
            @Override
            public void removeHandler() {
                addHandlers.remove(handler);
            }
        };
    }

    private void fireAdded(Widget w) {
        ContainerEvent event = new ContainerEvent(this, w);
        for (AddHandler h : addHandlers) {
            h.onAdd(event);
        }
    }

    private final Set<RemoveHandler> removeHandlers = new HashSet<>();

    @Override
    public HandlerRegistration addRemoveHandler(RemoveHandler handler) {
        removeHandlers.add(handler);
        return new HandlerRegistration() {
            @Override
            public void removeHandler() {
                removeHandlers.remove(handler);
            }
        };
    }

    private void fireRemoved(Widget w) {
        ContainerEvent event = new ContainerEvent(this, w);
        for (RemoveHandler h : removeHandlers) {
            h.onRemove(event);
        }
    }

    private JavaScriptObject componentAdded;
    
    public JavaScriptObject getComponentAdded() {
        return componentAdded;
    }

    protected HandlerRegistration componentAddedReg;

    public void setComponentAdded(JavaScriptObject aValue) {
        if (componentAdded != aValue) {
            if (componentAddedReg != null) {
                componentAddedReg.removeHandler();
                componentAddedReg = null;
            }
            componentAdded = aValue;
            if (componentAdded != null) {
                componentAddedReg = addAddHandler(new AddHandler() {
                    @Override
                    public void onAdd(ContainerEvent event) {
                        if (componentAdded != null) {
                            executeEvent(componentAdded, EventsPublisher.publishContainerEvent(event));
                        }
                    }
                });
            }
        }
    }

    private JavaScriptObject componentRemoved;
    
    public JavaScriptObject getComponentRemoved() {
        return componentRemoved;
    }

    protected HandlerRegistration componentRemovedReg;

    public void setComponentRemoved(JavaScriptObject aValue) {
        if (componentRemoved != aValue) {
            if (componentRemovedReg != null) {
                componentRemovedReg.removeHandler();
                componentRemovedReg = null;
            }
            componentRemoved = aValue;
            if (componentRemoved != null) {
                componentRemovedReg = addRemoveHandler(new RemoveHandler() {
                    @Override
                    public void onRemove(ContainerEvent event) {
                        if (componentRemoved != null) {
                            executeEvent(componentRemoved, EventsPublisher.publishContainerEvent(event));
                        }
                    }
                });
            }
        }
    }
}

File: platypus-js-web-client/src/app/kenga/src/container.js
define([
    'core/extend',
    './widget',
    './events/container-event',
    'core/invoke',
    'core/logger'], function (
        extend,
        Widget,
        ContainerEvent,
        Invoke,
        Logger) {
    function Container(shell, content) {
        if (!shell)
            shell = document.createElement('div');
        if (!content)
            content = shell;
        Widget.call(this, shell);
        var self = this;

        var children = [];

        this.element.classList.add('p-container');

        Object.defineProperty(this, 'count', {
            get: function () {
                return children.length;
            }
        });

        function child(index) {
            if (index >= 0 && index < children.length) {
                return children[index];
            } else {
                return null;
            }
        }
        Object.defineProperty(this, 'child', {
            configurable: true,
            get: function () {
                return child;
            }
        });
        function _children() {
            Logger.warning("'children()' function is obsolete. Use 'count', 'child' and 'forEach' please");
            return children.slice(0, children.length);
        }
        Object.defineProperty(this, 'children', {
            configurable: true,
            get: function () {
                return _children;
            }
        });

        function forEach(action) {
            children.forEach(action);
        }

        Object.defineProperty(this, 'forEach', {
            get: function () {
                return forEach;
            }
        });

        function indexOf(w) {
            return children.indexOf(w);
        }

        Object.defineProperty(this, 'indexOf', {
            get: function () {
                return indexOf;
            }
        });

        function add(w, beforeIndex) {
            if (w.parent === self)
                throw 'A widget is already child of this container';
            w.parent = self;
            if (isNaN(beforeIndex)) {
                content.appendChild(w.element);
                children.push(w);
            } else {
                if (beforeIndex < children.length) {
                    content.insertBefore(w.element, children[beforeIndex].element);
                } else {
                    content.appendChild(w.element);
                }
                children.splice(beforeIndex, 0, w);
            }
            fireAdded(w);
        }

        Object.defineProperty(this, 'add', {
            configurable: true,
            get: function () {
                return add;
            }
        });

        function remove(w) {
            var idx;
            if (w instanceof Widget)
                idx = children.indexOf(w);
            else
                idx = w;
            if (idx >= 0 && idx < children.length) {
                var removed = children.splice(idx, 1)[0];
                removed.parent = null;
                removed.element.parentElement.removeChild(removed.element);
                fireRemoved(w);
                return removed;
            } else {
                return null;
            }
        }

        Object.defineProperty(this, 'remove', {
            configurable: true,
            get: function () {
                return remove;
            }
        });

        function clear() {
            children.forEach(function (child) {
                child.parent = null;
                child.element.parentElement.removeChild(child.element);
                fireRemoved(child);
            });
            children.splice(0, children.length);
        }

        Object.defineProperty(this, 'clear', {
            configurable: true,
            get: function () {
                return clear;
            }
        });

        var addHandlers = new Set();
        function addAddHandler(handler) {
            addHandlers.add(handler);
            return {
                removeHandler: function () {
                    addHandlers.delete(handler);
                }
            };
        }

        Object.defineProperty(this, 'addAddHandler', {
            configurable: true,
            get: function () {
                return addAddHandler;
            }
        });

        function fireAdded(w) {
            var event = new ContainerEvent(self, w);
            addHandlers.forEach(function (h) {
                Invoke.later(function () {
                    h(event);
                });
            });
        }

        var onAdd;
        var addReg;
        Object.defineProperty(this, 'onAdd', {
            get: function () {
                return onAdd;
            },
            set: function (aValue) {
                if (onAdd !== aValue) {
                    if (addReg) {
                        addReg.removeHandler();
                        addReg = null;
                    }
                    onAdd = aValue;
                    if (onAdd) {
                        addReg = addAddHandler(function (event) {
                            if (onAdd) {
                                onAdd(event);
                            }
                        });
                    }
                }
            }
        });

        var removeHandlers = new Set();
        function addRemoveHandler(handler) {
            removeHandlers.add(handler);
            return {
                removeHandler: function () {
                    removeHandlers.delete(handler);
                }
            };
        }

        Object.defineProperty(this, 'addRemoveHandler', {
            configurable: true,
            get: function () {
                return addRemoveHandler;
            }
        });

        function fireRemoved(w) {
            var event = new ContainerEvent(self, w);
            removeHandlers.forEach(function (h) {
                Invoke.later(function () {
                    h(event);
                });
            });
        }

        var onRemove;
        var removeReg;
        Object.defineProperty(this, 'onRemove', {
            get: function () {
                return onRemove;
            },
            set: function (aValue) {
                if (onRemove !== aValue) {
                    if (removeReg) {
                        removeReg.removeHandler();
                        removeReg = null;
                    }
                    onRemove = aValue;
                    if (onRemove) {
                        removeReg = addRemoveHandler(function (event) {
                            if (onRemove) {
                                onRemove(event);
                            }
                        });
                    }
                }
            }
        });
    }
    extend(Container, Widget);
    return Container;
});